doc/api/diagnostics_channel.md

```diff
@@ -0,0 +1,282 @@
+# Diagnostics Channel
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+<!-- source_link=lib/diagnostics_channel.js -->
+
+The `diagnostics_channel` module provides an API to create named channels
+to report arbitrary message data for diagnostics purposes.
+
+It can be accessed using:
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+```
+
+It is intended that a module writer wanting to report diagnostics messages
+will create one or many top-level channels to report messages through.
+Channels may also be acquired at runtime but it is not encouraged
+due to the additional overhead of doing so. Channels may be exported for
+convenience, but as long as the name is known it can be acquired anywhere.
+
+If you intend for your module to produce diagnostics data for others to
+consume it is recommended that you include documentation of what named
+channels are used along with the shape of the message data.
+
+## Public API
+
+### Overview
+
+Following is a simple overview of the public API.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+// Get a reusable channel object
+const channel = diagnostics_channel.channel('my-channel');
+
+// Subscribe to the channel
+channel.subscribe((message, name) => {
+  // Received data
+});
+
+// Check if the channel has an active subscriber
+if (channel.shouldPublish()) {
+  // Publish data to the channel
+  channel.publish({
+    some: 'data'
+  });
+}
+
+// Spans correlate a sequence of events
+const span = channel.span({
+  some: 'start data'
+});
+
+// Annotations are for data not strictly related to the start or end events
+span.annotate({
+  some: 'annotation'
+});
+
+// This end event will share an `id` with the span started above
+span.end({
+  some: 'end data'
+});
+```
+
+#### `diagnostics_channel.shouldListen(name)`
```

Comment:
I think `hasSubscribers()` is a better name here. Seems also that the name in js files is actually `shouldPublish`.

Reply:
Oh, good catch. Not sure how I ended up with `shouldListen(...)` here. 😅 

Seems like a few people have asked for `hasSubscribers`, so I think I'll give that a try. What do you think on the function versus getter topic? There was another suggestion that it could be a getter.

Reply:
For the channel a getter is fine and easier to use.
I think that one on the module exports needs to be a function. But I think this top level function is not needed at all. if someone is concerned about performance a channel should be in it's hand.

Reply:
The reason for the top-level one is for the user to be able to check if there are channel subscribers without needing to create the channel to do it.